src/tests/test_upgrade_agent.py

```diff
@@ -0,0 +1,189 @@
+import datetime
+import json
+import os
+import re
+import subprocess
+from typing import List
+
+import kubernetes.client
+import kubernetes.config
+import kubernetes.dynamic
+from junit_report import JunitTestSuite
+
+from service_client import log
+from tests.base_test import BaseTest
+
+
+def skopeo(args: List[str]) -> str:
+    """
+    Runs the 'skopeo' command with the given command line arguments. Raises an exception if the command finishes with
+    a non zero exit code, or if it writes anything to the standard error stream. Returns an string containing whatever
+    the command wrote to the standard output stream.
+    """
+    process = subprocess.Popen(
+        args=["skopeo"] + args,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    (stdout, stderr) = process.communicate()
+    if process.returncode != 0:
+        raise Exception(f"failed to run skopeo with args '{args}', finished with code {process.returncode}")
+    if len(stderr) > 0:
+        raise Exception(f"failed to run skopeo with args '{args}', wrote '{stderr}' to the standard error stream")
+    return stdout.decode("utf-8")
+
+
+def image_repo(image: str) -> str:
+    """
+    Calculates the image repository from the given image reference. For example, if the image is
+    'quay.io/my/image:v1.23' it will return 'quay.io/my/image'.
+    """
+    return re.sub(r":.*$", "", image)
+
+
+def image_digest(image: str) -> str:
+    """
+    Gets the digest of the given image.
+    """
+    result = skopeo(["inspect", "--raw", f"docker://{image}"])
+    result = json.loads(result)
+    return result["config"]["digest"]
+
+
+def repo_tags(repo: str) -> str:
+    """
+    Gets the lit of tags of the given image respository"
+    """
+    result = skopeo(["list-tags", f"docker://{repo}"])
+    result = json.loads(result)
+    return result["Tags"]
+
+
+def parse_tag(tag: str) -> List[int]:
```

Comment:
we're usually using ``semver`` library for naturally sorting versions. you can take a look here: https://pypi.org/project/semver/

Reply:
wait. do we even have cases of ``v<x>.<y>.<z>-<r>``? Can you share with such an example? (because I don't remember such thing under ``quay.io/edge-infrastructure/*``)

either way, at least the parsing part (without using the compare functions) can be replaced so we'll maintain less code and less regex functions (and I think it's also okay to simplify the code to only consider simple, "no-r" releases)

Reply:
We don't have those cases in `quay.io/edge-infrastructure/*`, but we do in the downstream image repository, for example:

```
registry.redhat.io/rhai-tech-preview/assisted-installer-agent-rhel8:v1.0.0-187
```

That is what we are currently using in production. The fact that the first part is `1.0.0` and the second is large makes me think that we are never updating the first part, only the second. I this scenario if we ignore the second part then we could be potentially selecting a rather old version of the agent, maybe one that doesn't even have support for the upgrade feature. I don't know if we use these tests with that downstream repository. If we do then I think it is worth the extra complexity.

Reply:
But either way we cannot use this code to test staging/integration (and of course production) because all of them are to be managed via app-sre configuration and not by directly changing their configmaps

@lalon4 is it possible for the QE team to use this test logic for some of the use-cases?